### PR TITLE
Generic Metrics API

### PR DESCRIPTION
**Pull request by rmccorm4**
- [x] Test build variant with metrics disabled
- [x] Create unit test to validate new APIs
```
root@f862e5554bf3:/mnt/triton/jira/3284-metrics-api/core/build# LD_LIBRARY_PATH=$PWD/triton-core:$LD_LIBRARY_PATH ./triton-core/test/metrics_api_test 
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from MetricsApiTest
[ RUN      ] MetricsApiTest.TestCounterEndToEnd
# HELP api_counter_example this is an example counter metric added via API.
# TYPE api_counter_example counter
api_counter_example{example1="counter_label1",example2="counter_label2"} 10.000000

[       OK ] MetricsApiTest.TestCounterEndToEnd (2 ms)
[ RUN      ] MetricsApiTest.TestGaugeEndToEnd
# HELP api_gauge_example this is an example gauge metric added via API.
# TYPE api_gauge_example gauge
api_gauge_example{example1="gauge_label1",example2="gauge_label2"} 42.000000

[       OK ] MetricsApiTest.TestGaugeEndToEnd (1 ms)
[----------] 2 tests from MetricsApiTest (3 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (3 ms total)
[  PASSED  ] 2 tests.
```
- [x] Test calling actual `/metrics` endpoint equivalent (`Metrics::SerializedMetrics`) to view custom metrics
- [ ] Documentation - Separate PR to come with an example

**NOTE**

End-to-end testing from a sample backend will come with a separate PR, but does not necessarily have to block this PR given the unit tests added here.

### CODE REVIEW

**Comment by rmccorm4** [comment]
End-to-end testing from a sample backend will come with a separate PR, but does not necessarily have to block this PR given the unit tests added here.

File: include/triton/core/tritonserver.h
Comment (rmccorm4) [inline include/triton/core/tritonserver.h:94]: Please verify if this requires a version bump or not

Comment (tanmayv25) [reply include/triton/core/tritonserver.h:94]: Yes. Any change in this API should come with a version bump.

File: src/tritonserver.cc
Comment (rmccorm4) [inline src/tritonserver.cc:2742]: Need another set of eyes on this. 

Originally I retrieved the Registry object in the API implementation here:
```
const auto& registry = tc::Metrics::GetRegistry();
*family = reinterpret_cast<TRITONSERVER_MetricFamily*>(
      new tc::MetricFamily(kind, name, description, registry))
```

But the new metrics did not show up when collecting from the registry.

However, by passing the registry object from the API caller (unit test in this case), the new metrics were correctly added to the registry when calling `triton::core::Metrics::SerializedMetrics();`

File: include/triton/core/tritonserver.h
Comment (deadeyegoodwin) [inline include/triton/core/tritonserver.h:2211]: Increment/decrement only work for COUNTER? Set only works for GAUGE? If so document that and also the error code that is returned if called on incorrect kind metric.

**Review by deadeyegoodwin — CHANGES_REQUESTED** [review]

File: include/triton/core/tritonserver.h
Comment (rmccorm4) [reply include/triton/core/tritonserver.h:2169]: I agree, but I think I have a misunderstanding in how the static resource works, as getting the registry object in the API implementation did not end up capturing the new metrics. Would appreciate any ideas - see this comment: https://github.com/triton-inference-server/core/pull/46#discussion_r833841249

File: src/metric_family.cc
Comment (tanmayv25) [inline src/metric_family.cc:205]: Have a single return nullptr at the end and replace it with break in KIND_GAUGE case.

File: src/tritonserver.cc
Comment (tanmayv25) [inline src/tritonserver.cc:2744]: Use try-catch in conjunction with my above comments to return errors when the MetricFamilu creation failed.

File: src/tritonserver.cc
Comment (tanmayv25) [inline src/tritonserver.cc:2778]: try-catch here as well? Or may be not as it won't fail for a valid family with a kind.

File: src/metric_family.cc
Comment (tanmayv25) [inline src/metric_family.cc:228]: same as above

**Review by tanmayv25 — CHANGES_REQUESTED** [review]

File: include/triton/core/tritonserver.h
Comment (deadeyegoodwin) [reply include/triton/core/tritonserver.h:2170]: Did you enable the metrics? As is done by the server here: https://github.com/triton-inference-server/core/blob/main/src/tritonserver.cc#L1959-L1965

File: src/tritonserver.cc
Comment (rmccorm4) [reply src/tritonserver.cc]: Resolved by using C APIs to get metrics like `TRITONSERVER_ServerMetrics` and `TRITONSERVER_MetricsFormatted` instead of `tc::Metrics::SerializedMetrics()` in the unit test

File: include/triton/core/tritonserver.h
Comment (deadeyegoodwin) [inline include/triton/core/tritonserver.h:2226]: How is Decrement different than Increment with a negative value? Do we really need Decrement?

File: src/metric_family.cc
Comment (GuanLuo) [inline src/metric_family.cc:1]: `NVIDIA CORPORATION & AFFILIATES`

File: src/test/CMakeLists.txt
Comment (GuanLuo) [inline src/test/CMakeLists.txt:438]: If the unit test only interact with Triton core via C APIs, I don't think you need to include these files

File: include/triton/core/tritonserver.h
Comment (rmccorm4) [reply include/triton/core/tritonserver.h]: Good point, I will consolidate them.

File: include/triton/core/tritonserver.h
Comment (rmccorm4) [inline include/triton/core/tritonserver.h:2217]: @deadeyegoodwin there are some conflicting dictionary definitions of Increment for positive/negative values, though the mathematical definition doesn't require positive-ness. 

If you have any preferred name like Adjust/Change/etc. for clarity, let me know. I updated the documentation above the header to be more clear.

**Review by GuanLuo — DISMISSED** [review]
We should have a ticket to track upgrading to newer release of prometheus-cpp

File: src/metric_family.cc
Comment (rmccorm4) [inline src/metric_family.cc:196]: @deadeyegoodwin If we ever upgrade our prometheus-cpp dependency, this could be a few lines simpler and simply call `gauge_ptr->Increment(value)` regardless of sign. But I don't think we need to upgrade and risk potential changes to save 3 lines of code here.

Comment (rmccorm4) [reply src/metric_family.cc:196]: Similarly we could probably be fancy and do something like `gauge_ptr->Set(gauge_ptr->Value() + value);` (regardless of sign) here even with the current prometheus-cpp version, but I think that's less clear to read.

File: include/triton/core/tritonserver.h
Comment (deadeyegoodwin) [inline include/triton/core/tritonserver.h:2212]: Looks like you actually return INVALID_VALUE for a negative value with COUNTER

**Review by deadeyegoodwin — CHANGES_REQUESTED** [review]

File: include/triton/core/tritonserver.h
Comment (rmccorm4) [reply include/triton/core/tritonserver.h]: fixed

**Review by tanmayv25 — APPROVED** [review]

**Review by deadeyegoodwin — APPROVED** [review]

**Comment by rmccorm4** [comment]
✔️ Pipeline #4238580

**Pull request merged**
